File: packages/csharp/src/Models/Account.cs
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace Tesote.Sdk.Models;

/// <summary>Bank metadata embedded in an <see cref="Account"/>.</summary>
public sealed record AccountBank(
    [property: JsonPropertyName("name")] string Name);

/// <summary>Legal-entity attribution for an <see cref="Account"/>.</summary>
public sealed record AccountLegalEntity(
    [property: JsonPropertyName("id")] string? Id,
    [property: JsonPropertyName("legal_name")] string? LegalName);

/// <summary>Wire-side account-data envelope. Snake_case is preserved.</summary>
public sealed record AccountData(
    [property: JsonPropertyName("masked_account_number")] string MaskedAccountNumber,
    [property: JsonPropertyName("currency")] string Currency,
    [property: JsonPropertyName("transactions_data_current_as_of")] string? TransactionsDataCurrentAsOf,
    [property: JsonPropertyName("balance_data_current_as_of")] string? BalanceDataCurrentAsOf,
    [property: JsonPropertyName("custom_user_provided_identifier")] string? CustomUserProvidedIdentifier,
    [property: JsonPropertyName("balance_cents")] string? BalanceCents,
    [property: JsonPropertyName("available_balance_cents")] string? AvailableBalanceCents);

/// <summary>Account model. Identical between v1 and v2.</summary>
public sealed record Account(
    [property: JsonPropertyName("id")] string Id,
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("data")] AccountData Data,
    [property: JsonPropertyName("bank")] AccountBank Bank,
    [property: JsonPropertyName("legal_entity")] AccountLegalEntity LegalEntity,
    [property: JsonPropertyName("tesote_created_at")] string TesoteCreatedAt,
    [property: JsonPropertyName("tesote_updated_at")] string TesoteUpdatedAt);

/// <summary>Page-based pagination envelope (v1 / v2 accounts list).</summary>
public sealed record PageBasedPagination(
    [property: JsonPropertyName("current_page")] int CurrentPage,
    [property: JsonPropertyName("per_page")] int PerPage,
    [property: JsonPropertyName("total_pages")] int TotalPages,
    [property: JsonPropertyName("total_count")] int TotalCount);

/// <summary>Response envelope for GET /v1/accounts and /v2/accounts.</summary>
public sealed record AccountListResponse(
    [property: JsonPropertyName("total")] int Total,
    [property: JsonPropertyName("accounts")] IReadOnlyList<Account> Accounts,
    [property: JsonPropertyName("pagination")] PageBasedPagination Pagination);

File: packages/ts/src/models/account.ts
/**
 * Account model — identical between v1 and v2.
 * Wire shape uses snake_case; we preserve it on the model.
 */

export interface AccountBank {
  name: string;
}

export interface AccountLegalEntity {
  id: string | null;
  legal_name: string | null;
}

export interface AccountData {
  masked_account_number: string;
  currency: string;
  transactions_data_current_as_of: string | null;
  balance_data_current_as_of: string | null;
  custom_user_provided_identifier: string | null;
  /** Conditional: only present if `display_balances_in_api` is enabled. */
  balance_cents?: string;
  /** Conditional: only present if `display_balances_in_api` is enabled. */
  available_balance_cents?: string;
}

export interface Account {
  id: string;
  name: string;
  data: AccountData;
  bank: AccountBank;
  legal_entity: AccountLegalEntity;
  tesote_created_at: string;
  tesote_updated_at: string;
}

export interface PageBasedPagination {
  current_page: number;
  per_page: number;
  total_pages: number;
  total_count: number;
}

export interface AccountListResponse {
  total: number;
  accounts: Account[];
  pagination: PageBasedPagination;
}
